Flex time per day

dirkmb opened the pull request:
This PR fixes #47 'weekly/daily flexi time'

- Added a new event type called 'FLEX' which overrides the default FLEX-Time for this day.
- FLEX-Time is saved for each week in the database, so if the default flex time is changed over the time, the old weeks keep their flex time (until you re import the data)

My editor removed all trailing white spaces, sorry for theses many changes. Since I'm not a Java programmer I'm not sure if theses spaces where intentional. If you want to keep them let me know.

mathisdt commented on app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java:
```diff
@@ -119,15 +120,26 @@ private static void countTime(Map<Task, TimeSum> mapForCounting, Task task, Date
 	public DayLine calulateOneDay(DateTime day, List<Event> eventsOfOneDay) {
 		DayLine ret = new DayLine();
 
+        boolean found_day_flex_time = false;
```
```suggestion
        boolean foundDayFlexTime = false;
```

mathisdt commented on app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java:
```diff
@@ -119,15 +120,26 @@ private static void countTime(Map<Task, TimeSum> mapForCounting, Task task, Date
+            if (event.getType() == TypeEnum.FLEX.getValue()) {
+                DateTime flex_time = DateTimeUtil.stringToDateTime(event.getTime());
+                ret.getTimeFlexi().substract(flex_time.getHour(), flex_time.getMinute());
+                found_day_flex_time = true;
```
```suggestion
                foundDayFlexTime = true;
```

mathisdt commented on app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java:
```diff
@@ -119,15 +120,26 @@ private static void countTime(Map<Task, TimeSum> mapForCounting, Task task, Date
+        boolean found_day_flex_time = false;
+        for (Event event : eventsOfOneDay) {
+            if (event.getType() == TypeEnum.FLEX.getValue()) {
+                DateTime flex_time = DateTimeUtil.stringToDateTime(event.getTime());
```
```suggestion
                DateTime flexTime = DateTimeUtil.stringToDateTime(event.getTime());
```

mathisdt commented on app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java:
```diff
@@ -119,15 +120,26 @@ private static void countTime(Map<Task, TimeSum> mapForCounting, Task task, Date
+        for (Event event : eventsOfOneDay) {
+            if (event.getType() == TypeEnum.FLEX.getValue()) {
+                DateTime flex_time = DateTimeUtil.stringToDateTime(event.getTime());
+                ret.getTimeFlexi().substract(flex_time.getHour(), flex_time.getMinute());
```
```suggestion
                ret.getTimeFlexi().substract(flexTime.getHour(), flexTime.getMinute());
```

mathisdt commented on app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java:
```diff
@@ -119,15 +120,26 @@ private static void countTime(Map<Task, TimeSum> mapForCounting, Task task, Date
 		WeekDayEnum weekDay = WeekDayEnum.getByValue(day.getWeekDay());
 		if (Basics.getInstance().getPreferences().getBoolean(Key.ENABLE_FLEXI_TIME.getName(), false)
-			&& timerManager.isWorkDay(weekDay)) {
+			&& timerManager.isWorkDay(weekDay) && found_day_flex_time == false) {
```
```suggestion
			&& timerManager.isWorkDay(weekDay) && foundDayFlexTime == false) {
```

mathisdt commented on app/src/main/java/org/zephyrsoft/trackworktime/timer/TimeCalculator.java:
```diff
@@ -119,15 +120,26 @@ private static void countTime(Map<Task, TimeSum> mapForCounting, Task task, Date
 		}
 
-		if (eventsOfOneDay == null || eventsOfOneDay.isEmpty()) {
+		if (eventsOfOneDay == null || eventsOfOneDay.isEmpty() || found_day_flex_time && eventsOfOneDay.size() == 1) {
```
```suggestion
		if (eventsOfOneDay == null || eventsOfOneDay.isEmpty() || foundDayFlexTime && eventsOfOneDay.size() == 1) {
```

mathisdt commented on app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java:
```diff
@@ -308,11 +315,51 @@ public TimeSum calculateTimeSum(DateTime date, PeriodEnum periodEnum) {
+		for (WeekDayEnum weekDay : WeekDayEnum.values()) {
+			DateTime day = date.plusDays(weekDay.getValue() - 1);
+			events = dao.getEventsOnDay(day);
+			boolean found_flex_time = false;
```
```suggestion
			boolean foundFlexTime = false;
```

mathisdt commented on app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java:
```diff
@@ -308,11 +315,51 @@ public TimeSum calculateTimeSum(DateTime date, PeriodEnum periodEnum) {
+				DateTime flex_time = DateTimeUtil.stringToDateTime(event.getTime());
+				ret.add(eventTime.getHour(), eventTime.getMinute());
+				Logger.debug("calculating flexi time for {} {} event found: {}", day.toString(), weekDay.getValue(), flex_time.toString());
+				found_flex_time = true;
```
```suggestion
				foundFlexTime = true;
```

mathisdt commented on app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java:
```diff
@@ -308,11 +315,51 @@ public TimeSum calculateTimeSum(DateTime date, PeriodEnum periodEnum) {
+				found_flex_time = true;
+				break;  // there can only be one flext time per day
+			}
+			if (found_flex_time == false) {
```
```suggestion
			if (foundFlexTime == false) {
```

mathisdt commented on app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java:
```diff
@@ -308,11 +315,51 @@ public TimeSum calculateTimeSum(DateTime date, PeriodEnum periodEnum) {
+				ret.add(eventTime.getHour(), eventTime.getMinute());
+				Logger.debug("calculating flexi time for {} {} event found: {}", day.toString(), weekDay.getValue(), flex_time.toString());
+				found_flex_time = true;
+				break;  // there can only be one flext time per day
```
```suggestion
				break;  // there can only be one flexi time per day
```

mathisdt commented on app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java:
```diff
@@ -604,13 +658,19 @@ public void updateWeekSum(final Week week) {
 			weekToUse.setSum(0);
 		}
+
+		TimeSum flexi_sum = calculateFlexTimeSum(DateTimeUtil.stringToDateTime(week.getStart()));
```
```suggestion
		TimeSum flexiSum = calculateFlexTimeSum(DateTimeUtil.stringToDateTime(week.getStart()));
```

mathisdt commented on app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java:
```diff
@@ -604,13 +658,19 @@ public void updateWeekSum(final Week week) {
 		}
+
+		TimeSum flexi_sum = calculateFlexTimeSum(DateTimeUtil.stringToDateTime(week.getStart()));
+		int flexi_minutes = flexi_sum.getAsMinutes();
```
```suggestion
		int flexiMinutes = flexiSum.getAsMinutes();
```

mathisdt commented on app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java:
```diff
@@ -604,13 +658,19 @@ public void updateWeekSum(final Week week) {
+		TimeSum flexi_sum = calculateFlexTimeSum(DateTimeUtil.stringToDateTime(week.getStart()));
+		int flexi_minutes = flexi_sum.getAsMinutes();
+		Logger.info("updating the flexi time sum to {} minutes for the week beginning at {}", flexi_minutes, week.getStart());
+		weekToUse.setFlexi(flexi_minutes);
```
```suggestion
		weekToUse.setFlexi(flexiMinutes);
```

mathisdt commented on app/src/main/java/org/zephyrsoft/trackworktime/timer/TimerManager.java:
```diff
@@ -604,13 +658,19 @@ public void updateWeekSum(final Week week) {
+
+		TimeSum flexi_sum = calculateFlexTimeSum(DateTimeUtil.stringToDateTime(week.getStart()));
+		int flexi_minutes = flexi_sum.getAsMinutes();
+		Logger.info("updating the flexi time sum to {} minutes for the week beginning at {}", flexi_minutes, week.getStart());
```
```suggestion
		Logger.info("updating the flexi time sum to {} minutes for the week beginning at {}", flexiMinutes, week.getStart());
```

mathisdt commented:
Next steps: when the variable name issue is resolved, I will merge the PR and run the new dev version on my own device for a while before creating an official new version of the app. But I won't disregard it - it just may take some time as I do this on my (rare) spare time.

dirkmb commented:
hi, thx for the fast and detailed review. I changed all variable names and updated the MR.
Sorry it took so long, like you I also had only very rare spare time.

mathisdt commented:
I'm not sure I have understood what the new feature is doing.

1. I have the following situation without using a flex event. No work was tracked on Thursday so flexi time is reduced.
![events-without-flex](https://user-images.githubusercontent.com/1932462/64991935-257a9380-d8d3-11e9-85b8-e60f67b022a2.png)
![overview-without-flex](https://user-images.githubusercontent.com/1932462/64991873-febc5d00-d8d2-11e9-8f08-aadbeb9ff416.png)

2. Now I add a flex event on Thursday and have the following:
![events-with-flex](https://user-images.githubusercontent.com/1932462/64991976-3aefbd80-d8d3-11e9-851e-44110a8cca59.png)
![overview-with-flex](https://user-images.githubusercontent.com/1932462/64991965-375c3680-d8d3-11e9-9eda-165069166c87.png)

As far as I can see, the flex event prohibits that flexi time is reduced on a day on which no work was tracked (Thursday in the above example). Is that what the flex event should do?

dirkmb commented:
I guess the use of 'FLEX' as event name confused you, or it does not say what it is doing. This event defines the needed/wanted flex time for the given day. So instead of 8 hours you are supposed to work only 0 hours on Thursday.

You have already 9 hours on your flexi-time (overtime). After you configured the FLEX event for Thursday you needed to work 0 hours for this day. So you do have the previous 9 hours of flexi-time on you clock.
If you would work 1 hour on Thursday you would have 10 hours flexi-time.

Does this explain it?
